src/protocols/aux_gen.rs

```diff
@@ -222,63 +131,138 @@ impl<P: SchemeParams, Id: PartyId> ProtocolError<Id> for AuxGenError<P, Id> {
+    ) -> Result<(), EvidenceError> {
+        let sid = Sid::new::<P, Id>(shared_randomness, &shared_data.ids);
+
+        match &self.error {
+            R3ErrorEnum::ModFailed => {
+                let rid = reconstruct_rid::<P, _>(&messages)?;
                 let aux = (&sid, guilty_party, &rid);
-                let r2_bc = previous_messages
-                    .get_round(2)?
-                    .normal_broadcast
-                    .deserialize::<Round2NormalBroadcast<P>>(format)?;
-                let r3_bc = message
-                    .normal_broadcast
-                    .deserialize::<Round3NormalBroadcast<P>>(format)?;
-                let paillier_pk = r2_bc.paillier_pk.into_precomputed();
-                verify_that(!r3_bc.psi_prime.verify(&paillier_pk, &aux))
+                let r2_nb = messages.previous_normal_broadcast::<Round2<P, Id>>(2)?;
+                let r3_nb = messages.normal_broadcast()?;
+                let paillier_pk = r2_nb.paillier_pk.into_precomputed();
+                verify_that(!r3_nb.psi_prime.verify(&paillier_pk, &aux))
             }
-            Error::R3FacFailed { reported_by } => {
-                let rid = reconstruct_rid::<P, _>(format, &previous_messages, &combined_echos)?;
+            R3ErrorEnum::FacFailed { reported_by } => {
+                let rid = reconstruct_rid::<P, _>(&messages)?;
                 let aux = (&sid, guilty_party, &rid);
 
-                let r2_eb = combined_echos
-                    .get_round(2)?
-                    .try_get("combined echos for Round 2", reported_by)?
-                    .deserialize::<Round2EchoBroadcast<P>>(format)?;
-                let r2_bc = previous_messages
-                    .get_round(2)?
-                    .normal_broadcast
-                    .deserialize::<Round2NormalBroadcast<P>>(format)?;
-                let r3_dm = message.direct_message.deserialize::<Round3DirectMessage<P>>(format)?;
-                let paillier_pk = r2_bc.paillier_pk.into_precomputed();
+                let r2_eb = messages
+                    .combined_echos::<Round2<P, Id>>(2)?
+                    .get_or_invalid_evidence("combined echos for Round 2", reported_by)?
+                    .clone();
+                let r2_nb = messages.previous_normal_broadcast::<Round2<P, Id>>(2)?;
+                let r3_dm = messages.direct_message()?;
+                let paillier_pk = r2_nb.paillier_pk.into_precomputed();
                 let rp_params = r2_eb.rp_params.to_precomputed();
                 verify_that(!r3_dm.psi.verify(&paillier_pk, &rp_params, &aux))
             }
         }
     }
 }
 
+/// Reconstruct `rid` from echoed messages
+fn reconstruct_rid<P: SchemeParams, Id: PartyId>(
+    messages: &EvidenceMessages<'_, Id, Round3<P, Id>>,
+) -> Result<BitVec, EvidenceError> {
+    let r2_ebs = messages.combined_echos::<Round2<P, Id>>(2)?;
+    let r2_eb = messages.previous_echo_broadcast::<Round2<P, Id>>(2)?;
```

Comment:
We discussed this a fair bit during the `manul` review, but I nonetheless was surprised to see that we have to pass in the `2` here. It's a wart. :/

Reply:
Yeah I understand it, but still had to pause. It's just a bit of an oddity. Maybe if rounds had "functional" names it'd be less jarring? Like, dunno, "KeyInitRound", "ReshareRound", etc